File: split/finalign.pas
procedure finalign;
var
  p, q, r, s, u, v: halfword;
  t, w: scaled;
  o: scaled;
  n: halfword;
  rulesave: scaled;
  auxsave: memoryword;
begin
  if curgroup <> 6 then
    confusion(914);
  unsave;
  if curgroup <> 6 then
    confusion(915);
  unsave;
  if nest[nestptr - 1].modefield = 203 then
    o := eqtb[5845].int
  else
    o := 0;
  {801:}
q := mem[mem[29992].hh.rh].hh.rh;
  repeat
    flushlist(mem[q + 3].int);
    flushlist(mem[q + 2].int);
    p := mem[mem[q].hh.rh].hh.rh;
    if mem[q + 1].int = -1073741824 then{802:}

    begin
      mem[q + 1].int := 0;
      r := mem[q].hh.rh;
      s := mem[r + 1].hh.lh;
      if s <> 0 then
      begin
        mem[0].hh.rh := mem[0].hh.rh + 1;
        deleteglueref(s);
        mem[r + 1].hh.lh := 0;
      end;
    end{:802}
;
    if mem[q].hh.lh <> 29991 then{803:}

    begin
      t := mem[q + 1].int + mem[mem[mem[q].hh.rh + 1].hh.lh + 1].int;
      r := mem[q].hh.lh;
      s := 29991;
      mem[s].hh.lh := p;
      n := 1;
      repeat
        mem[r + 1].int := mem[r + 1].int - t;
        u := mem[r].hh.lh;
        while mem[r].hh.rh > n do
        begin
          s := mem[s].hh.lh;
          n := mem[mem[s].hh.lh].hh.rh + 1;
        end;
        if mem[r].hh.rh < n then
        begin
          mem[r].hh.lh := mem[s].hh.lh;
          mem[s].hh.lh := r;
          mem[r].hh.rh := mem[r].hh.rh - 1;
          s := r;
        end
        else
        begin
          if mem[r + 1].int > mem[mem[s].hh.lh + 1].int then
            mem[mem[s].hh.lh + 1].int := mem[r + 1].int;
          freenode(r, 2);
        end;
        r := u;
      until r = 29991;
    end{:803}
;
    mem[q].hh.b0 := 13;
    mem[q].hh.b1 := 0;
    mem[q + 3].int := 0;
    mem[q + 2].int := 0;
    mem[q + 5].hh.b1 := 0;
    mem[q + 5].hh.b0 := 0;
    mem[q + 6].int := 0;
    mem[q + 4].int := 0;
    q := p;
  until q = 0{:801}
;
  {804:}
saveptr := saveptr - 2;
  packbeginline := -curlist.mlfield;
  if curlist.modefield = -1 then
  begin
    rulesave := eqtb[5846].int;
    eqtb[5846].int := 0;
    p := hpack(mem[29992].hh.rh, savestack[saveptr + 1].int, savestack[saveptr + 0].int);
    eqtb[5846].int := rulesave;
  end
  else
  begin
    q := mem[mem[29992].hh.rh].hh.rh;
    repeat
      mem[q + 3].int := mem[q + 1].int;
      mem[q + 1].int := 0;
      q := mem[mem[q].hh.rh].hh.rh;
    until q = 0;
    p := vpackage(mem[29992].hh.rh, savestack[saveptr + 1].int, savestack[saveptr + 0].int, 1073741823);
    q := mem[mem[29992].hh.rh].hh.rh;
    repeat
      mem[q + 1].int := mem[q + 3].int;
      mem[q + 3].int := 0;
      q := mem[mem[q].hh.rh].hh.rh;
    until q = 0;
  end;
  packbeginline := 0{:804}
;
  {805:}
q := mem[curlist.headfield].hh.rh;
  s := curlist.headfield;
  while q <> 0 do
  begin
    if not (q >= himemmin) then
      if mem[q].hh.b0 = 13 then
        {807:}
begin
        if curlist.modefield = -1 then
        begin
          mem[q].hh.b0 := 0;
          mem[q + 1].int := mem[p + 1].int;
        end
        else
        begin
          mem[q].hh.b0 := 1;
          mem[q + 3].int := mem[p + 3].int;
        end;
        mem[q + 5].hh.b1 := mem[p + 5].hh.b1;
        mem[q + 5].hh.b0 := mem[p + 5].hh.b0;
        mem[q + 6].gr := mem[p + 6].gr;
        mem[q + 4].int := o;
        r := mem[mem[q + 5].hh.rh].hh.rh;
        s := mem[mem[p + 5].hh.rh].hh.rh;
        repeat
          {808:}
n := mem[r].hh.b1;
          t := mem[s + 1].int;
          w := t;
          u := 29996;
          while n > 0 do
          begin
            n := n - 1;
            {809:}
s := mem[s].hh.rh;
            v := mem[s + 1].hh.lh;
            mem[u].hh.rh := newglue(v);
            u := mem[u].hh.rh;
            mem[u].hh.b1 := 12;
            t := t + mem[v + 1].int;
            if mem[p + 5].hh.b0 = 1 then
            begin
              if mem[v].hh.b0 = mem[p + 5].hh.b1 then
                t := t + round(mem[p + 6].gr * mem[v + 2].int);
            end
            else if mem[p + 5].hh.b0 = 2 then
              if mem[v].hh.b1 = mem[p + 5].hh.b1 then
                t := t - round(mem[p + 6].gr * mem[v + 3].int);
            s := mem[s].hh.rh;
            mem[u].hh.rh := newnullbox;
            u := mem[u].hh.rh;
            t := t + mem[s + 1].int;
            if curlist.modefield = -1 then
              mem[u + 1].int := mem[s + 1].int
            else
            begin
              mem[u].hh.b0 := 1;
              mem[u + 3].int := mem[s + 1].int;
            end{:809}
;
          end;
          if curlist.modefield = -1 then{810:}

          begin
            mem[r + 3].int := mem[q + 3].int;
            mem[r + 2].int := mem[q + 2].int;
            if t = mem[r + 1].int then
            begin
              mem[r + 5].hh.b0 := 0;
              mem[r + 5].hh.b1 := 0;
              mem[r + 6].gr := 0.0;
            end
            else if t > mem[r + 1].int then
            begin
              mem[r + 5].hh.b0 := 1;
              if mem[r + 6].int = 0 then
                mem[r + 6].gr := 0.0
              else
                mem[r + 6].gr := (t - mem[r + 1].int) / mem[r + 6].int;
            end
            else
            begin
              mem[r + 5].hh.b1 := mem[r + 5].hh.b0;
              mem[r + 5].hh.b0 := 2;
              if mem[r + 4].int = 0 then
                mem[r + 6].gr := 0.0
              else if (mem[r + 5].hh.b1 = 0) and (mem[r + 1].int - t > mem[r + 4].int) then
                mem[r + 6].gr := 1.0
              else
                mem[r + 6].gr := (mem[r + 1].int - t) / mem[r + 4].int;
            end;
            mem[r + 1].int := w;
            mem[r].hh.b0 := 0;
          end{:810}

          else{811:}

          begin
            mem[r + 1].int := mem[q + 1].int;
            if t = mem[r + 3].int then
            begin
              mem[r + 5].hh.b0 := 0;
              mem[r + 5].hh.b1 := 0;
              mem[r + 6].gr := 0.0;
            end
            else if t > mem[r + 3].int then
            begin
              mem[r + 5].hh.b0 := 1;
              if mem[r + 6].int = 0 then
                mem[r + 6].gr := 0.0
              else
                mem[r + 6].gr := (t - mem[r + 3].int) / mem[r + 6].int;
            end
            else
            begin
              mem[r + 5].hh.b1 := mem[r + 5].hh.b0;
              mem[r + 5].hh.b0 := 2;
              if mem[r + 4].int = 0 then
                mem[r + 6].gr := 0.0
              else if (mem[r + 5].hh.b1 = 0) and (mem[r + 3].int - t > mem[r + 4].int) then
                mem[r + 6].gr := 1.0
              else
                mem[r + 6].gr := (mem[r + 3].int - t) / mem[r + 4].int;
            end;
            mem[r + 3].int := w;
            mem[r].hh.b0 := 1;
          end{:811}
;
          mem[r + 4].int := 0;
          if u <> 29996 then
          begin
            mem[u].hh.rh := mem[r].hh.rh;
            mem[r].hh.rh := mem[29996].hh.rh;
            r := u;
          end{:808}
;
          r := mem[mem[r].hh.rh].hh.rh;
          s := mem[mem[s].hh.rh].hh.rh;
        until r = 0;
      end{:807}

      else if mem[q].hh.b0 = 2 then{806:}

      begin
        if (mem[q + 1].int = -1073741824) then
          mem[q + 1].int := mem[p + 1].int;
        if (mem[q + 3].int = -1073741824) then
          mem[q + 3].int := mem[p + 3].int;
        if (mem[q + 2].int = -1073741824) then
          mem[q + 2].int := mem[p + 2].int;
        if o <> 0 then
        begin
          r := mem[q].hh.rh;
          mem[q].hh.rh := 0;
          q := hpack(q, 0, 1);
          mem[q + 4].int := o;
          mem[q].hh.rh := r;
          mem[s].hh.rh := q;
        end;
      end{:806}
;
    s := q;
    q := mem[q].hh.rh;
  end{:805}
;
  flushnodelist(p);
  popalignment;
  {812:}
auxsave := curlist.auxfield;
  p := mem[curlist.headfield].hh.rh;
  q := curlist.tailfield;
  popnest;
  if curlist.modefield = 203 then{1206:}

  begin
    doassignments;
    if curcmd <> 3 then{1207:}

    begin
      begin
        if interaction = 3 then;
        printnl(262);
        print(1169);
      end;
      begin
        helpptr := 2;
        helpline[1] := 894;
        helpline[0] := 895;
      end;
      backerror;
    end{:1207}

    else{1197:}

    begin
      getxtoken;
      if curcmd <> 3 then
      begin
        begin
          if interaction = 3 then;
          printnl(262);
          print(1165);
        end;
        begin
          helpptr := 2;
          helpline[1] := 1166;
          helpline[0] := 1167;
        end;
        backerror;
      end;
    end{:1197}
;
    popnest;
    begin
      mem[curlist.tailfield].hh.rh := newpenalty(eqtb[5274].int);
      curlist.tailfield := mem[curlist.tailfield].hh.rh;
    end;
    begin
      mem[curlist.tailfield].hh.rh := newparamglue(3);
      curlist.tailfield := mem[curlist.tailfield].hh.rh;
    end;
    mem[curlist.tailfield].hh.rh := p;
    if p <> 0 then
      curlist.tailfield := q;
    begin
      mem[curlist.tailfield].hh.rh := newpenalty(eqtb[5275].int);
      curlist.tailfield := mem[curlist.tailfield].hh.rh;
    end;
    begin
      mem[curlist.tailfield].hh.rh := newparamglue(4);
      curlist.tailfield := mem[curlist.tailfield].hh.rh;
    end;
    curlist.auxfield.int := auxsave.int;
    resumeafterdisplay;
  end{:1206}

  else
  begin
    curlist.auxfield := auxsave;
    mem[curlist.tailfield].hh.rh := p;
    if p <> 0 then
      curlist.tailfield := q;
    if curlist.modefield = 1 then
      buildpage;
  end{:812}
;
end;

File: cpp/finalign.cpp
#include "finalign.h"
#include "confusion.h"
#include "unsave.h"
#include "flushlist.h"
#include "deleteglueref.h"
#include "freenode.h"
#include "hpack.h"
#include "vpackage.h"
#include "newglue.h"
#include "newnullbox.h"
#include "flushnodelist.h"
#include "popalignment.h"
#include "popnest.h"
#include "printnl.h"
#include "doassignments.h"
#include "backerror.h"
#include "print.h"
#include "getxtoken.h"
#include "resumeafterdisplay.h"
#include "buildpage.h"
#include "newpenalty.h"
#include "newparamglue.h"
#include <cmath>
#include "texte.h"

void finalign(void)
{
	halfword p, q, r, s, u, v;
	scaled t, w;
	scaled o;
	halfword n;
	scaled rulesave;
	memoryword auxsave;
	if (curgroup != 6)
		confusion("align1");
	unsave();
	if (curgroup != 6)
		confusion("align0"); 
	unsave();
	if (nest[nestptr-1].modefield == mmode)
		o = display_indent();
	else
		o = 0;
	q = link(preamble());
	do
	{
		flushlist(u_part(q));
		flushlist(v_part(q));
		p = link(link(q));
		if (is_running(width(q)))
		{
			width(q) = 0;
			r = link(q);
			s = info(r+1);
			if (s)
			{
				link(zero_glue)++;
				deleteglueref(s);
				info(r+1) = 0;
			}
		}
		if (info(q) != end_span)
		{
			t = width(q)+width(info(link(q)+1));
			r = info(q);
			s = end_span;
			info(s) = p;
			n = 1;
			do
			{
				width(r) -= t;
				u = info(r);
				while (link(r) > n)
				{
					s = info(s);
					n = link(info(s))+1;
				}
				if (link(r) < n)
				{
					info(r) = info(s);
					info(s) = r;
					link(r)--;
					s = r;
				}
				else
				{
					if (width(r) > width(info(s)+1))
						width(info(s)+1) = width(r);
					freenode(r, span_node_size);
				}
				r = u;
			} while (r != end_span);
		}
		type(q) = unset_node;
		span_count(q) = 0;
		height(q) = 0;
		depth(q) = 0;
		glue_order(q) = 0;
		glue_sign(q) = 0;
		glue_stretch(q) = 0;
		glue_shrink(q) = 0;
		q = p;
	} while (q);
	saveptr -= 2;
	packbeginline = -mode_line;
	if (mode == -vmode)
	{
		rulesave = overfull_rule();
		overfull_rule() = 0;
		p = hpack(preamble(), savestack[saveptr+1].int_, savestack[saveptr].int_);
		overfull_rule() = rulesave;
	}
	else
	{
		q = link(preamble());
		do
		{
			height(q) = width(q);
			width(q) = 0;
			q = link(link(q));
		} while (q);
		p = vpack(preamble(), savestack[saveptr+1].int_, savestack[saveptr+0].int_);
		q = link(preamble());
		do
		{
			width(q) = height(q);
			height(q) = 0;
			q = link(link(q));
		} while (q);
	}
	packbeginline = 0;
	q = link(head);
	s = head;
	while (q)
	{
		if (q < himemmin)
			if (type(q) == unset_node)
			{
				if (mode == -vmode)
				{
					type(q) = hlist_node;
					width(q) = width(p);
				}
				else
				{
					type(q) = vlist_node;
					height(q) = height(p);
				}
				glue_order(q) = glue_order(p);
				glue_sign(q) = glue_sign(p);
				glue_set(q) = glue_set(p);
				shift_amount(q) = o;
				r = link(list_ptr(q));
				s = link(list_ptr(p));
				do
				{
					n = span_count(r);
					t = width(s);
					w = t;
					u = hold_head;
					while (n > 0)
					{
						n--;
						s = link(s);
						v = info(s+1);
						link(u) = newglue(v);
						u = link(u);
						subtype(u) = 12;
						t += width(v);
						if (glue_sign(p) == stretching)
						{
							if (type(v) == glue_order(p))
								t += round(glue_set(p)*depth(v));
						}
						else 
							if (glue_sign(p) == shrinking && subtype(v) == glue_order(p))
								t -= round(glue_set(p)*height(v));
						s = link(s);
						link(u) = newnullbox();
						u = link(u);
						t += width(s);
						if (mode == -vmode)
							width(u) = width(s);
						else
						{
							type(u) = vlist_node;
							height(u) = height(s);
						}
					}
					if (mode == -vmode)
					{
						height(r) = height(q);
						depth(r) = depth(q);
						if (t == width(r))
						{
							glue_sign(r) = 0;
							glue_order(r) = 0;
							glue_set(r) = 0.0;
						}
						else 
							if (t > width(r))
							{
								glue_sign(r) = stretching;
								if (glue_stretch(r) == 0)
									glue_set(r) = 0.0;
								else
									glue_set(r) = (t-width(r))/glue_stretch(r);
							}
							else
							{
								glue_order(r) = glue_sign(r);
								glue_sign(r) = shrinking;
								if (glue_shrink(r) == 0)
									glue_set(r) = 0.0;
								else 
									if (glue_order(r) == 0 && width(r)-t > glue_shrink(r))
										glue_set(r) = 1.0;
									else
										glue_set(r) = (width(r)-t) / glue_shrink(r);
							}
						width(r) = w;
						type(r) = 0;
					}
					else
					{
						width(r) = width(q);
						if (t == height(r))
						{
							glue_sign(r) = 0;
							glue_order(r) = 0;
							glue_set(r) = 0.0;
						}
						else 
							if (t > height(r))
							{
								glue_sign(r) = stretching;
								if (glue_stretch(r) == 0)
									glue_set(r) = 0.0;
								else
									glue_set(r) = (t-height(r)) / glue_stretch(r);
							}
							else
							{
								glue_order(r) = glue_sign(r);
								glue_sign(r) = shrinking;
								if (glue_shrink(r) == 0)
									glue_set(r) = 0.0;
								else 
									if (glue_order(r) == 0 && height(r)-t > glue_shrink(r))
										glue_set(r) = 1.0;
									else
										glue_set(r) = (height(r)-t) / glue_shrink(r);
							}
						height(r) = w;
						type(r) = vlist_node;
					}
					shift_amount(r) = 0;
					if (u != hold_head)
					{
						link(u) = link(r);
						link(r) = link(hold_head);
						r = u;
					}
					r = link(link(r));
					s = link(link(s));
				} while (r);
			}
			else 
				if (type(q) == rule_node)
				{
					if (is_running(width(q)))
						width(q) = width(p);
					if (is_running(height(q)))
						height(q) = height(p);
					if (is_running(depth(q)))
						depth(q) = depth(p);
					if (o)
					{
						r = link(q);
						link(q) = 0;
						q = hpack(q, 0, additional);
						shift_amount(q) = o;
						link(q) = r;
						link(s) = q;
					}
				}
		s = q;
		q = link(q);
	}
	flushnodelist(p);
	popalignment();
	auxsave = aux;
	p = link(head);
	q = tail;
	popnest();
	if (mode == mmode)
	{
		doassignments();
		if (curcmd != math_shift)
		{
			printnl("! ");
			print("Missing $$ inserted"); 
			helpptr = 2;
			helpline[1] = "Displays can use special alignments (like \\eqalignno)";
			helpline[0] = "only if nothing but the alignment itself is between $$'s.";
			backerror();
		}
		else
		{
			getxtoken();
			if (curcmd != math_shift)
			{
				printnl("! ");
				print("Display math should end with $$");
				helpptr = 2;
				helpline[1] = "The `$' that I just saw supposedly matches a previous `$$'.";
				helpline[0] = "So I shall assume that you typed `$$' both times.";
				backerror();
			}
		}
		popnest();
		tail_append(newpenalty(pre_display_penalty()));
		tail_append(newparamglue(3));
		link(tail) = p;
		if (p)
			tail = q;
		tail_append(newpenalty(post_display_penalty()));
		tail_append(newparamglue(4));
		aux = auxsave;
		resumeafterdisplay();
	}
	else
	{
		aux = auxsave;
		link(tail) = p;
		if (p)
			tail = q;
		if (mode == vmode)
			buildpage();
	}
}
